OSDOCS-7663: Added a new Quick Start Guide that uses Terraform (#64339)
openshift/openshift-docs: modules/rosa-sts-cluster-terraform-customization.adoc

// Module included in the following assemblies:
//
// * rosa_install_access_delete_clusters/rosa-sts-creating-a-cluster-quickly-terraform.adoc
//
ifeval::["{context}" == "rosa-sts-creating-a-cluster-quickly-terraform"]
:tf-defaults:
endif::[]
:_content-type: PROCEDURE

= Creating your ROSA cluster with Terraform

> The following Terraform cluster example creates your account-wide IAM roles then creates a ROSA cluster with a managed OIDC configuration.

Reviewer: ```suggestion
The following Terraform-cluster example shows you how to create account-wide IAM roles and a ROSA cluster with a managed OIDC configuration.
```

.Procedure

> . Optional: since the Terraform files that are created during this process are created in your current directory, you can create a new directory to store these files. Run the following command to create and navigate into a new directory:

Reviewer: ```suggestion
. Optional: Because the Terraform files are created in your current directory during this procedure, you can create a new directory to store these files. 
.. To create and navigate into a new directory, run the following command:
```
I would use "because" here. "Since" might translate as a time word ("...since you've been gone...")
There's surely a better link for this, but we cap after the colon of Optional steps (footnote #2) https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#callouts-in-code-blocks 
I prefer the reverse construction for giving the "run the command" instruction because it is easier scan down the list for the one I want if the thing I am doing is at the beginning of the sentence.

+
[source,terminal]
----
$ mkdir terraform-cluster && cd terraform-cluster
----

> . You can grant permissions to your account by using link:https://console.redhat.com/openshift/token[an offline {cluster-manager-first} token]. You set the token as a variable by running the following command:

Reviewer: ```suggestion
. You can grant permissions to your account by using link:https://console.redhat.com/openshift/token[an offline {cluster-manager-first} token]. 
.. Set the token as a variable by running the following command:
```

+
[source,terminal]
----
$ export RHCS_TOKEN=<your_offline_token>
----
+

> After you have exported your token, you can verify this value by running the following command:

Reviewer: ```suggestion
After you export your token, verify the value by running the following command:
```

+
[source,terminal]
----
$ echo $RHCS_TOKEN
----

. Create the Terraform files locally by using the following code templates:

> .. Run the following command to create the `account-roles.tf` file:

Reviewer: ```suggestion
.. Create the `account-roles.tf` file by running the following command:
```

+
[source,terminal]
----

> $ cat<<-EOF>account-roles.tf

Reviewer: see note below in the standard assembly about checking these blocks after pub

data "rhcs_policies" "all_policies" {}

data "rhcs_versions" "all" {}

module "create_account_roles" {
  source  = "terraform-redhat/rosa-sts/aws"
  version = "0.0.15"

  create_account_roles  = true
  create_operator_roles = false

  account_role_prefix    = local.cluster_name
  path                   = var.path
  rosa_openshift_version = regex("^[0-9]+\\\\.[0-9]+", var.rosa_openshift_version)
  account_role_policies  = data.rhcs_policies.all_policies.account_role_policies
  all_versions           = data.rhcs_versions.all
  operator_role_policies = data.rhcs_policies.all_policies.operator_role_policies
  tags                   = var.additional_tags
}

resource "time_sleep" "wait_10_seconds" {
  depends_on = [module.create_account_roles]

  create_duration = "10s"
}
EOF
----

> .. Run the following command to create the `main.tf` file:

Reviewer: ```suggestion
.. Create the `main.tf` file by running the following command:
```

endif::tf-defaults[]
ifndef::tf-defaults[]
  autoscaling_enabled  = var.autoscaling_enabled
  min_replicas         = var.min_replicas
  max_replicas         = var.max_replicas
endif::tf-defaults[]
  sts                  = local.sts_roles
  properties = {
    rosa_creator_arn = data.aws_caller_identity.current.arn
  }
  machine_cidr     = var.vpc_cidr_block
ifndef::tf-defaults[]
  #Private link settings

  private          = var.private_cluster
  aws_private_link = var.private_cluster
  aws_subnet_ids   = var.create_vpc ? concat(module.vpc[0].private_subnets, module.vpc[0].public_subnets) : var.private_subnet_ids
  machine_cidr     = var.private_cluster ? var.vpc_cidr_block : null
endif::tf-defaults[]

  lifecycle {
    precondition {
      condition     = can(regex("^[a-z][-a-z0-9]{0,13}[a-z0-9]\$", local.cluster_name))
      error_message = "ROSA cluster name must be less than 16 characters, be lower case alphanumeric, with only hyphens."
    }
  }

  depends_on = [time_sleep.wait_10_seconds]
}

resource "rhcs_cluster_wait" "wait_for_cluster_build" {
  cluster = rhcs_cluster_rosa_classic.rosa_sts_cluster.id
  # timeout in minutes
  timeout = 60
}
EOF
----
endif::tf-defaults[]

> .. Run the following command to create the `oidc-provider.tf` file:

Reviewer: ```suggestion
.. Create the `oidc-provider.tf` file by running the following command:
```

+
[source,terminal]
----
$ cat<<-EOF>oidc-provider.tf
resource "rhcs_rosa_oidc_config" "oidc_config" {
  managed = true
}

data "rhcs_rosa_operator_roles" "operator_roles" {
  operator_role_prefix = local.cluster_name
  account_role_prefix  = local.cluster_name
}

module "oidc_provider" {
  source  = "terraform-redhat/rosa-sts/aws"
  version = "0.0.15"

  create_operator_roles = false
  create_oidc_provider  = true

  cluster_id                  = ""
  rh_oidc_provider_thumbprint = rhcs_rosa_oidc_config.oidc_config.thumbprint
  rh_oidc_provider_url        = rhcs_rosa_oidc_config.oidc_config.oidc_endpoint_url
  tags                        = var.additional_tags
  path                        = var.path
}
EOF
----

> .. Run the following command to create the `variables.tf` file:

Reviewer: ```suggestion
.. Create the `variables.tf` file by running the following command:
```

  default     = false
}

#VPC Info
variable "create_vpc" {
  default     = false
  type        = bool
  description = "Would you like to make a new VPC for your ROSA cluster? true or false"
}

variable "vpc_name" {
  type        = string
  description = "VPC Name"
  default     = "rosa-tf-vpc"
}

variable "vpc_cidr_block" {
  type        = string
  description = "The value of the IP address block for machines or cluster nodes for the VPC."
  default     = "10.0.0.0/16"
}

variable "private_subnet_cidrs" {
  type        = list(any)
  description = "The CIDR blocks to use for the private subnets"
}

variable "public_subnet_cidrs" {
  type        = list(any)
  description = "The CIDR blocks to use for the public subnets"
  default     = ["10.0.101.0/24", "10.0.102.0/24", "10.0.103.0/24"]
}

variable "single_nat_gateway" {
  type        = bool
  description = "Single NAT or per NAT for subnet"
  default     = false
}

#AWS Info
variable "aws_region" {
  type    = string
  default = "us-east-1"
}

variable "private_subnet_ids" {
  type        = list(any)
  description = "VPC private subnets IDs for ROSA Cluster"
  default     = []
}
EOF
----

> .. Run the following command to create the `vpc.tf` file:

Reviewer: ```suggestion
.. Create the `vpc.tf` file by running the following command:
```

+
[source,terminal]
----
$ cat<<-EOF>vpc.tf
module "vpc" {
  source  = "terraform-aws-modules/vpc/aws"
  version = "5.1.2"

  count = var.create_vpc ? 1 : 0
  name  = var.vpc_name
  cidr  = var.vpc_cidr_block

  azs             = local.region_azs
  private_subnets = var.private_subnet_cidrs
  public_subnets  = var.public_subnet_cidrs

  enable_nat_gateway   = true
  single_nat_gateway   = var.single_nat_gateway
  enable_dns_hostnames = true
  enable_dns_support   = true

  tags = var.additional_tags
}
EOF
----
endif::tf-defaults[]
. Create a `terraform.tfvars` file by running the following command:
+
[source,terminal]
----
$ cat<<-EOF>terraform.tfvars
# Advanced ROSA Cluster Options
worker_node_replicas = "2"

#Private Link Cluster & VPC
single_nat_gateway=true
ifndef::tf-defaults[]
vpc_cidr_block="10.12.0.0/16"
endif::tf-defaults[]
private_subnet_cidrs=["10.12.0.0/22", "10.12.4.0/22", "10.12.8.0/22"]
public_subnet_cidrs=["10.12.128.0/24", "10.12.129.0/24", "10.12.130.0/24"]
EOF
----

> . In your terminal, run the following command in the directory where you saved the Terraform files to set up Terraform to create these resources:

Reviewer: ```suggestion
. Start up a working directory containing the Terraform configuration files by running the following command:
```
Aren't we already in the terminal? And shouldn't we already be in the relevant directory whether we completed Step 1 or not (because the Terraform files go to the current directory no matter what, we said in Step 1).

If this is a common breaking step because the user doesn't do this process all at once, an Important admonition might be better; something like, "You must complete the following steps from the directory where you created the Terraform files."

+
[source,terminal]
----
$ terraform init
----

> . Optional: Run the following command to confirm that the Terraform you copied is correct:

Reviewer: ```suggestion
. Optional: Verify that the Terraform you copied is correct by running the following command:
```
"copied"? or is it "configured"?

+
[source,terminal]
----
$ terraform validate
----
+

> .Sample output

Reviewer: ```suggestion
.Example output
```

+
[source,terminal]
----
Success! The configuration is valid.
----

> . Run the following command to build your cluster with Terraform:

Reviewer: ```suggestion
. Build your cluster with Terraform by running the following command:
```
"Build?" I'm not familiar with that term for clusters. Does this start the pods? Or is this a configuration/staging step?

+
[source,terminal]
----
$ terraform apply
----

> . When prompted, enter a name for your cluster.

Reviewer: . The Terraform interface lists the resources to be created or changed and prompts for confirmation. Enter `yes` to proceed, or `no` to cancel:
```

+
[source,terminal]
----
var.cluster_name
  Provide the name of your ROSA cluster.

>   Enter a value:

Reviewer: ```suggestion
  Enter a value: <yes_or_no> <1>
```
Then you can drop a footnote like <1> Valid values are `yes` or `no`.

----
+

> After entering a cluster name, the Terraform plan shows what is created or changed. Enter `yes` to start the role and cluster creation.

Reviewer: ```suggestion
If you enter `yes`, your Terraform plan executes, creating your AWS account roles, Operator roles, and your ROSA Classic cluster.
```
Entering the cluster name should be another step.

+

> [IMPORTANT]

Reviewer: Do we need this note in the other procedure as well? I didn't see it in the diff, not sure its there.

====
Your cluster name must be fewer than 16 characters.
====

.Verification

> . Run the following command to verify that your cluster was created:

Reviewer: ```suggestion
. Verify that your cluster was created by running the following command:
```

+
[source,terminal]
----
$ rosa list clusters
----

> . Your cluster's ID, name, and status are shown in the terminal:

Reviewer: ```suggestion

```

+

> .Sample output

Reviewer: ```suggestion
.Example output showing the cluster ID, name, and status
```

[source,terminal]
----
ID                                NAME          STATE  TOPOLOGY
27c3snjsupa9obua74ba8se5kcj11269  rosa-tf-demo  ready  Classic (STS)
----

> . Run the following command to verify that your account roles were created:

Reviewer: ```suggestion
. Verify that your account roles were created by running the following command:
```

+
[source,terminal]
----
$ rosa list account-roles
----

> . The Terraform-created account roles are shown.

Reviewer: ```suggestion

```

+

> .Sample output

Reviewer: ```suggestion
.Example output showing Terraform-created account roles
```

[source,terminal]
----
I: Fetching account roles
ROLE NAME                                   ROLE TYPE      ROLE ARN                                                           OPENSHIFT VERSION  AWS Managed
ROSA-demo-ControlPlane-Role                 Control plane  arn:aws:iam::<ID>:role/ROSA-demo-ControlPlane-Role                 4.13               No
ROSA-demo-Installer-Role                    Installer      arn:aws:iam::<ID>:role/ROSA-demo-Installer-Role                    4.13               No
ROSA-demo-Support-Role                      Support        arn:aws:iam::<ID>:role/ROSA-demo-Support-Role                      4.13               No
ROSA-demo-Worker-Role                       Worker         arn:aws:iam::<ID>:role/ROSA-demo-Worker-Role                       4.13               No
----

> . Run the following command to verify that your Operator roles were created:

Reviewer: ```suggestion
. Verify that your Operator roles were created by running the following command:
```

+
[source,terminal]
----
$ rosa list operator-roles
----

> . The Terraform-created Operator roles are shown.

Reviewer: ```suggestion

```

+

> .Sample output

Reviewer: ```suggestion
.Example output listing Terraform-created Operator roles
```
